File: matrix_algorithms/src/matrices/Algoritmo3_NaivLoopUnrollingFour.java
package matrices;

import java.math.BigInteger;

// Algoritmo 3: NaivLoopUnrollingFour
// Complejidad: O(n^3) con overhead de ciclo reducido al 25%
// Desenrolla el ciclo interno de a 4 pasos por iteracion
// Procesa 4 productos simultaneamente por iteracion del ciclo k
// El residuo (p % 4 elementos) se procesa con un ciclo simple al final
public class Algoritmo3_NaivLoopUnrollingFour {

    public static BigInteger[][] multiplicar(BigInteger[][] A, BigInteger[][] B) {
        int n = A.length, m = B[0].length, p = B.length;
        BigInteger[][] C = new BigInteger[n][m];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++) {
                BigInteger aux = BigInteger.ZERO;
                int k = 0;
                // procesa de a 4 elementos por iteracion
                for (; k < p - 3; k += 4)
                    aux = aux.add(A[i][k].multiply(B[k][j]))
                             .add(A[i][k+1].multiply(B[k+1][j]))
                             .add(A[i][k+2].multiply(B[k+2][j]))
                             .add(A[i][k+3].multiply(B[k+3][j]));
                // elementos restantes (0, 1, 2 o 3 segun p % 4)
                for (; k < p; k++)
                    aux = aux.add(A[i][k].multiply(B[k][j]));
                C[i][j] = aux;
            }
        return C;
    }
}

File: matrix_algorithms_python/algoritmo3_naiv_loop_unrolling_four.py
# Algoritmo 3: NaivLoopUnrollingFour
# Complejidad: O(n^3) con overhead de ciclo reducido al 25%
# Desenrolla el ciclo interno de a 4 pasos por iteracion
# El residuo (n % 4 elementos) se procesa con un ciclo simple al final
def multiplicar(A, B):
    n = len(A)
    C = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            aux = 0
            k = 0
            # procesa de a 4 elementos por iteracion
            while k < n - 3:
                aux += (A[i][k]   * B[k][j]   +
                        A[i][k+1] * B[k+1][j] +
                        A[i][k+2] * B[k+2][j] +
                        A[i][k+3] * B[k+3][j])
                k += 4
            # elementos restantes (0, 1, 2 o 3 segun n % 4)
            while k < n:
                aux += A[i][k] * B[k][j]
                k += 1
            C[i][j] = aux
    return C
